accounts-on-demand/src/accounts_on_demand.rs

```diff
@@ -0,0 +1,228 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use dashmap::DashSet;
+use itertools::Itertools;
+use solana_client::{
+    nonblocking::rpc_client::RpcClient,
+    rpc_config::{RpcAccountInfoConfig, RpcProgramAccountsConfig},
+    rpc_filter::RpcFilterType,
+};
+use solana_lite_rpc_accounts::account_store_interface::AccountStorageInterface;
+use solana_lite_rpc_cluster_endpoints::geyser_grpc_connector::GrpcSourceConfig;
+use solana_lite_rpc_core::{
+    commitment_utils::Commitment,
+    structures::{
+        account_data::{AccountData, AccountNotificationMessage},
+        account_filter::{AccountFilter, AccountFilterType, AccountFilters},
+    },
+};
+use solana_sdk::{clock::Slot, pubkey::Pubkey};
+use tokio::sync::{broadcast::Sender, RwLock};
+
+use crate::subscription_manager::SubscriptionManger;
+
+pub struct AccountsOnDemand {
+    rpc_client: Arc<RpcClient>,
+    accounts_storage: Arc<dyn AccountStorageInterface>,
+    accounts_subscribed: Arc<DashSet<Pubkey>>,
+    program_filters: Arc<RwLock<AccountFilters>>,
+    subscription_manager: SubscriptionManger,
+}
+
+impl AccountsOnDemand {
+    pub fn new(
+        rpc_client: Arc<RpcClient>,
+        grpc_sources: Vec<GrpcSourceConfig>,
+        accounts_storage: Arc<dyn AccountStorageInterface>,
+        account_notification_sender: Sender<AccountNotificationMessage>,
+    ) -> Self {
+        Self {
+            rpc_client,
+            accounts_storage: accounts_storage.clone(),
+            accounts_subscribed: Arc::new(DashSet::new()),
+            program_filters: Arc::new(RwLock::new(vec![])),
+            subscription_manager: SubscriptionManger::new(
+                grpc_sources,
+                accounts_storage,
+                account_notification_sender,
+            ),
+        }
+    }
+
+    pub async fn reset_subscription(&self) {
+        let mut filters = self.get_filters().await;
+
+        // add additional filters related to accounts
+        for accounts in &self
+            .accounts_subscribed
+            .iter()
+            .map(|x| x.to_string())
+            .chunks(100)
+        {
+            let account_filter = AccountFilter {
+                accounts: accounts.collect(),
+                program_id: None,
+                filters: None,
+            };
+            filters.push(account_filter);
+        }
+
+        self.subscription_manager
+            .update_subscriptions(filters)
+            .await;
+    }
+
+    pub async fn get_filters(&self) -> AccountFilters {
+        self.program_filters.read().await.clone()
+    }
+}
+
+#[async_trait]
+impl AccountStorageInterface for AccountsOnDemand {
+    async fn update_account(&self, account_data: AccountData, commitment: Commitment) -> bool {
+        self.accounts_storage
+            .update_account(account_data, commitment)
+            .await
+    }
+
+    async fn initilize_account(&self, account_data: AccountData) {
```

Comment:
typo

Reply:
It is defined in the trait which aims to warmup the cache.